Project Sorting and Bug Squashing!

rickbau5 opened the pull request:
Dual purpose pull, some bug squashing and a few minor & major features.

Major:
- Project list view now only has 16 projects per page, and can scroll through multiple pages (#118)
- Project list view can now be sorted with the dropdown menu. Available sorting is by title, author, and dates. This is compatible with the pagination.
- Refactored `ProjectsTableSeeder` to use a new `MassSeeder` class. This class enables you to do multiple iterations of the seeding functionality while getting a pretty loading bar to look at. It can be configured to accept CLI input, or from the config, and works with Travis CI
- Added SafeDeleteProjects command (`php artisan projects:Delete {num}` num=number, or "all"). This provides a way to drop some amount of projects from the database, and also delete their image. Dropping tables through MySQL/PHPMyAdmin doesn't delete images.

Bug Fixes:
- Fixed the cancel button on create file page, and updated the cancel button on create project page. Both now use `back()` instead of a redirect.
- Fixed the display location of the delete notification banner.

@uidaho/lambda Ready for review.

rickbau5 committed:
Make test executable

Now can just do ./test from directory and run the tests. Merely a helper

rickbau5 committed:
Fix Cancel buttons

Cancel buttons on file and project create pages now redirect back to the last page. Fixes #109

rickbau5 committed:
Enable error message on reset page

rickbau5 committed:
Fix delete notification position

Resolves #113

rickbau5 committed:
Implement sorting on projects page

Resolves #116

rickbau5 committed:
Modify query building

This way will preserve other queries in the url. For instance if there was already `?page=1` the previous way would overwrite that.

rickbau5 committed:
Add Projects table mass seeder

- Now there is a MassSeeder class that, if extended, will all easy duplication of a seeding behavior.
- Make mass seeder play nice with travis

rickbau5 committed:
Implement Project sorting and pagination

- Projects are now limited to 16 per page
- Projects can be sorted by Title, author, and dates.

Fixes #116 and #118

rickbau5 committed:
Add default alert div to main_layout

Doing Session::flash('alert', 'My message') will now show an alert just under the nav bar for every page

rickbau5 committed:
Bug fixes

brec9824 commented:
Hey do you get this error if you clear out your database, migrate, then run db:seed
![capture](https://cloud.githubusercontent.com/assets/16866862/14731694/1c2f8014-0808-11e6-858b-9c36821d439b.PNG)

rickbau5 commented:
Hmm no? I dunno why that's not working D:
![image](https://cloud.githubusercontent.com/assets/11299264/14731732/8c13dce0-0808-11e6-901b-d87b7e2b3ab2.png)

brec9824 commented:
@rickbau5 Had to run composer update.

rickbau5 commented:
oh huh ok. I don't really get all of what that even does still lol

brec9824 commented on database/seeds/ProjectsTableMassSeeder.php:
```diff
@@ -0,0 +1,41 @@
+    {
+        factory(Project::class, 1)->create([
+            'author' => $this->faker->userName,
+            'title' => $this->faker->realText(32),
```
The faker should return something else because sometimes it returns symbols that breaks the url.

rickbau5 replied:
:+1: didn't think of that

brec9824 commented on database/seeds/MassSeeder.php:
```diff
@@ -0,0 +1,105 @@
+        $this->setup();
+
+        $times = $this->getNumberIterations();
+        $bar = $this->command->getOutput()->createProgressBar($times);
```
Nice!

rickbau5 replied:
So nice, especially since this seeder takes a while. Even for 5!

brec9824 commented on app/Http/Controllers/ProjectController.php:
```diff
@@ -5,18 +5,49 @@
+    public function listProjects()
+    {
+        $allProjects = Project::all();
+        $sorting = Request::has('sort');
```
I dont understand how you are sorting but there seems to be a problem. Because it seems to have problems with sorting between words with capitols and words with lowercase. How are you sorting?

brec9824 commented:
Seems when you switch pages the sort goes crazy and you have to click on the type of sort you want again. The urls go from /projects/?sort=title&page=2 to /projects?sort=1&page=3 and if I change the 1 in the url to title its fine again.

rickbau5 commented on app/Http/Controllers/ProjectController.php:
```diff
@@ -5,18 +5,49 @@
+        $sorting = Request::has('sort');
+
+        if ($sorting) {
+            $allProjects = $allProjects->sortBy(Request::get('sort'));
```
@brec9824 Sorting happens here, it uses the built in Model's method of sorting so it's beyond our control.

rickbau5 committed:
Add SafeDeleteCommand

- New command to safe delete many projects, thereby deleting there image too. Do `php artisan projects:delete {num}` where {num} is a number or "all".
- Removed hold command PopulateProjectsTable, done by seeder now.

rickbau5 committed:
Fix sorting, and fix punctuation in titles

rickbau5 committed:
Untrack duplicate images and add .gitignore

- .gitignore added to preserve directory structure

rickbau5 commented:
@brec9824 fixed the sorting bug

rickbau5 commented:
@brec9824 all fixed, and squashed a bit.

brec9824 commented:
Modified date sort appears to be wrong. Not sure what order it is. Sort by author was wrong.
What it was giving me at the end:
-gGrady
-gHand
-tReynolds
-uMurazik
-uRuecker
-Lew.Schmidt
-zStanton

Phpmyadmin:
-uRuecker
-Volkman.Layne
-vStreich
-Winston.Glover
-zStanton
-Zulauf.Braden

I think the symbols, capitals and whatnot is giving it problems.
Sort by title worked with a few that were off by one or two spots. The created_at worked.

rickbau5 commented:
@brec9824 strange i'm seeing that too, I hadn't noticed before: 
![image](https://cloud.githubusercontent.com/assets/11299264/14734608/75cd15ce-0820-11e6-9ed1-0b820df13d0d.png)
8 from the bottom it's wrong again. hmmmm.

rickbau5 commented:
I should've looked at your comment closer, it's totally the upper vs lower case stuff haha. Working on it

rickbau5 committed:
Fix sorting bug with case

rickbau5 commented:
@brec9824 alright fixed, I checked and it looks good. I was putting this after `$allProjects = $allProjects->sort...` to see the output:
 `dd(array_map(function($project) { return $project[Request::get('sort')]; }, $allProjects->toArray()));`

brec9824 commented:
@rickbau5 Looks good now. Only thing I have to mention is it would be nice to have sorting by 
-author a->z
-author z->a
-title a->z
-title z->a
-newest->oldest (probably created at)
-oldest->newest (probably created at)

If not let me know and I will merge.

rickbau5 committed:
Add ascending/descending sort mode

rickbau5 commented:
@brec9824 done. I think i need to add some visual indication of sort mode but i couldn't come up with something at the moment. I'll work on it a bit more today.

brec9824 commented:
It might be worth looking into replacing the current drop down menu with one that displays the current sort taking place.

rickbau5 committed:
Tweak display of sorting

rickbau5 commented:
@brec9824 did that, but now that i'm looking at it. I really should make a custom request for this. So much crap in the controller. I'll do that, hold off on merging.

rickbau5 committed:
Abstract away sorting

rickbau5 committed:
Abstract away Pagination logic from controller

New type of Request: PaginatedRequest. See ProjectListRequest for example implementation

rickbau5 committed:
Add documentation

brec9824 commented:
@rickbau5 is this all good for review or are you still adding stuff.

rickbau5 commented:
@brec9824 good for revie

brec9824 commented on app/Http/Requests/SortableRequest.php:
```diff
@@ -0,0 +1,68 @@
+
+    public function getSortOrder()
+    {
+        switch ($this->get('order'))
```
Not really a problem but why not an if else for such a small check?

Commit:
Remove empty link on sort boxes

Removed the links on the sort name and the ascending or descending boxes. Previously was redirecting to the index page.
